### PR TITLE
fix(compute_ctl): Allow usage of DB names with whitespaces

### PR DESCRIPTION
**Pull request by ololobus**
## Problem

We used `set_path()` to replace the database name in the connection string. It automatically does url-safe encoding if the path is not already encoded, but it does it as per the URL standard, which assumes that tabs can be safely removed from the path without changing the meaning of the URL. See, e.g., https://url.spec.whatwg.org/#concept-basic-url-parser. It also breaks for DBs with properly %-encoded names, like with `%20`, as they are kept intact, but actually should be escaped.

Yet, this is not true for Postgres, where it's completely valid to have trailing tabs in the database name.

I think this is the PR that caused this regression https://github.com/neondatabase/neon/pull/9717, as it switched from `postgres::config::Config` back to `set_path()`.

This was fixed a while ago already [1], btw, I just haven't added a test to catch this regression back then :(

## Summary of changes

This commit changes the code back to use `postgres/tokio_postgres::Config` everywhere.

While on it, also do some changes around, as I had to touch this code:
1. Bump some logging from `debug` to `info` in the spec apply path. We do not use `debug` in prod, and it was tricky to understand what was going on with this bug in prod.
2. Refactor configuration concurrency calculation code so it was reusable. Yet, still keep `1` in the case of reconfiguration. The database can be actively used at this moment, so we cannot guarantee that there will be enough spare connection slots, and the underlying code won't handle connection errors properly.
3. Simplify the installed extensions code. It was spawning a blocking task inside async function, which doesn't make much sense. Instead, just have a main sync function and call it with `spawn_blocking` in the API code -- the only place we need it to be async.
4. Add regression python test to cover this and related problems in the future. Also, add more extensive testing of schema dump and DBs and roles listing API.

[1]: https://github.com/neondatabase/neon/commit/4d1e48f3b9a4b7064787513fd2c455f0001f6e18
[2]: https://www.postgresql.org/message-id/flat/20151023003445.931.91267%40wrigleys.postgresql.org

Resolves neondatabase/cloud#20869

### CODE REVIEW

**Comment by ololobus** [comment]
I wonder, is the way I did it this time is also OK, because I'm not sure how to switch to `postgres::config::Config` in all places now. Schema dump is the most problematic, as it just passes connstr to the cli

File: compute_tools/src/compute.rs [hunk 1]
```diff
@@ -1222,21 +1224,24 @@ impl ComputeNode {
...
-                url.query_pairs_mut()
-                    .append_pair("application_name", "apply_config");
-                let url = Arc::new(url);
+                let mut conf = tokio_postgres::Config::from_str(self.connstr.as_str()).unwrap();
+                conf.application_name("apply_config");
+                let conf = Arc::new(conf);
 
                 let spec = Arc::new(spec.clone());
 
-                self.apply_spec_sql(spec, url, 1)?;
```
Comment (ololobus) [inline compute_tools/src/compute.rs:1239]: I think this `1` just meant that we still processed reconfiguration sequentially. And for projects with 1k DBs that'd timeout, compute will be recreated and stuck again. Refactored this to use the same logic. @MMeent thoughts?

File: compute_tools/src/compute.rs [hunk 2]
```diff
@@ -1061,34 +1063,16 @@ impl ComputeNode {
...
-            compute_state
-                .pspec
-                .as_ref()
-                .expect("spec must be set")
-                .spec
-                .clone(),
-        );
-
-        // Choose how many concurrent connections to use for applying the spec changes.
-        // If the cluster is not currently Running we don't have to deal with user connections,
```
Comment (ololobus) [inline compute_tools/src/compute.rs:1088]: We actually never reach this code in Running state, it's either of
1. Init -- initial startup and configuration
2. Configuration -- when we reconfigure the running compute

So changed to Init, because I think that was the original idea

File: compute_tools/src/compute.rs [hunk 3]
```diff
@@ -964,14 +966,14 @@ impl ComputeNode {
                     match &db {
                         DB::SystemDB => {}
                         DB::UserDB(db) => {
-                            url.set_path(db.name.as_str());
+                            conf.dbname(db.name.as_str());
                         }
                     }
 
-                    let url = Arc::new(url);
+                    let conf = Arc::new(conf);
```
Comment (ololobus) [inline compute_tools/src/compute.rs:973]: This code with a bunch of Arcs and clones looked a bit awkward to me even before my changes, but I haven't found a good way to change it quickly

**Comment by tristan957** [comment]
is it possible to add a test?

File: compute_tools/src/pg_helpers.rs [hunk 4]
```diff
@@ -542,3 +545,30 @@ async fn handle_postgres_logs_async(stderr: tokio::process::ChildStderr) -> Resu
...
+/// it does it as per the URL standard, which assumes that tabs can be safely
+/// removed from the path without changing the meaning of the URL. See, e.g.,
+/// <https://url.spec.whatwg.org/#concept-basic-url-parser>.
+///
+/// Yet, this is not true for Postgres, where it's completely valid to have
+/// trailing tabs in the database name. Here, we first encode the database name
+/// and then swap it in the connection string.
+///
+/// XXX: this is still not fully correct, and the only correct way is to use
+/// `conn_config_for_db()` instead.
```
Comment (hlinnaka) [inline compute_tools/src/pg_helpers.rs:561]: In what cases does it go wrong?

Comment (hlinnaka) [reply compute_tools/src/pg_helpers.rs:561]: I think you mean `postgres_conf_for_db` here

File: compute_tools/src/spec.rs [hunk 5]
```diff
@@ -1,15 +1,14 @@
...
 use std::fs::File;
 use std::path::Path;
 use tracing::{error, info, instrument, warn};
 
-use crate::config;
-use crate::migration::MigrationRunner;
-use crate::params::PG_HBA_ALL_MD5;
-use crate::pg_helpers::*;
-
 use compute_api::responses::{ControlPlaneComputeStatus, ControlPlaneSpecResponse};
```
Comment (hlinnaka) [inline compute_tools/src/spec.rs:12]: unnecessary change

**Review by hlinnaka — APPROVED** [review]
Looks good at high level. Please add test and address all the minor comments though

File: compute_tools/src/installed_extensions.rs [hunk 6]
```diff
@@ -42,75 +41,51 @@ fn list_dbs(client: &mut Client) -> Result<Vec<String>> {
 ///
 /// Same extension can be installed in multiple databases with different versions,
 /// we only keep the highest and lowest version across all databases.
-pub async fn get_installed_extensions(connstr: Url) -> Result<InstalledExtensions> {
```
Comment (hlinnaka) [inline compute_tools/src/installed_extensions.rs:45]: This is just incidental refactoring, right? Maybe split off to a separate PR

**Comment by hlinnaka** [comment]
> is it possible to add a test?

+1. The `connstr_for_db` function would be an ideal candidate for a unit test. Maybe `max_service_connections` too.

**Comment by hlinnaka** [comment]
Let's put out the fire first, but let's also consider if this is actually an upstream bug in tokio-postgres, and submit a fix there too.

**Comment by ololobus** [comment]
> +1. The connstr_for_db function would be an ideal candidate for a unit test. Maybe max_service_connections too.

Added a python test instead for now. It fails in main and passes with my changes. There are many places, and I'm afraid that with some future refactoring someone will use a set_path() again in some of the places. +1 for writing more unit tests and doing some further refactoring, but will keep it for further work

File: compute_tools/src/installed_extensions.rs (hunk 6 above)
Comment (ololobus) [reply compute_tools/src/installed_extensions.rs:45]: Yeah, but it was so hard to read, so I decided to just do it as well. It's a bit hard to separate now, so let's keep it altogether

**Comment by ololobus** [comment]
Going to merge when CI is ready

**Comment by hlinnaka** [comment]
> > Let's put out the fire first, but let's also consider if this is actually an upstream bug in tokio-postgres, and submit a fix there too.
> 
> @hlinnaka What do you mean by upstream bug in `tokio-postgres`? We used `set_path()`, which is a part of Url crate. I think it actually obeys the standard. After switching back to postgres/tokio-postgres::Config usage, it works just fine

Oh right, I somehow thought this was in tokio-postgres. Never mind

**Comment by hlinnaka** [comment]
`
        let base_url = url::Url::parse("postgres://user@example.com/basedb").unwrap();

        connstr_for_db(&base_url, "foo db").as_str()
```
returns `postgres://user@example.com/foo+db`. But at least psql interprets that as a + sign, not as space

File: compute_tools/src/compute.rs (hunk 1 above)
Comment (ololobus) [reply compute_tools/src/compute.rs:1239]: Reverted this back to 1, but here is the image that was built with higher concurrency. We can test it on the project with 10k DBs

Compute with higher apply config concurrency `neondatabase/vm-compute-node-v16:12071102747`
https://github.com/neondatabase/neon/actions/runs/12071102747/job/33663022735?pr=9919#step:8:1

File: test_runner/regress/test_compute_catalog.py [hunk 7]
```diff
@@ -33,3 +33,63 @@ def test_compute_catalog(neon_simple_env: NeonEnv):
...
+            "owner": "cloud_admin",
+        },
+        {
+            "name": "db with%20spaces ",
+            "owner": "cloud_admin",
+        },
+        {
+            "name": "db with whitespaces	",
+            "owner": "cloud_admin",
+        },
```
[the remark this answers is not in the document]
Comment (ololobus) [reply test_runner/regress/test_compute_catalog.py]: Added even more tricky and injective cases

File: compute_tools/src/pg_helpers.rs [hunk 8]
```diff
@@ -542,3 +545,34 @@ async fn handle_postgres_logs_async(stderr: tokio::process::ChildStderr) -> Resu
...
+///
+/// Yet, this is not true for Postgres, where it's completely valid to have
+/// trailing tabs in the database name. Here, we first encode the database name
+/// and then swap it in the connection string.
+pub fn connstr_for_db(connstr: &url::Url, dbname: &str) -> url::Url {
+    let mut connstr = connstr.clone();
+    let encoded: String = url::form_urlencoded::byte_serialize(dbname.as_bytes()).collect();
+    connstr.set_path(&encoded);
+    connstr
+}
```
[the remark this answers is not in the document]
Comment (ololobus) [reply compute_tools/src/pg_helpers.rs]: Refactored schema dump + extended test, so dropped this helper altogether. It was flawed anyway

File: test_runner/regress/test_compute_catalog.py [hunk 9]
```diff
@@ -33,3 +33,63 @@ def test_compute_catalog(neon_simple_env: NeonEnv):
...
+                "databases": databases,
+            },
+        }
+    )
+    endpoint.reconfigure()
+
+    for db in databases:
+        # Check that database has a correct name in the system catalog
+        catalog_db = endpoint.safe_psql(
+            f"SELECT datname FROM pg_database WHERE datname = '{db['name']}'"
```
[the remark this answers is not in the document]
Comment (ololobus) [reply test_runner/regress/test_compute_catalog.py]: Added proper escaping

**Comment by ololobus** [comment]
> returns postgres://user@example.com/foo+db. But at least psql interprets that as a + sign, not as space

It's still bad. I think I've finally got rid of messing with urls everywhere, so just dropped this helper

File: compute_tools/src/compute.rs [hunk 10]
```diff
@@ -931,7 +932,7 @@ impl ComputeNode {
                 RenameAndDeleteDatabases,
                 CreateAndAlterDatabases,
             ] {
-                debug!("Applying phase {:?}", &phase);
+                info!("Applying phase {:?}", &phase);
```
[the remark this answers is not in the document]
Comment (ololobus) [reply compute_tools/src/compute.rs:935]: Same for both logging comments -- it's currently not clear from logs: what's the current phase if it takes too long and what was the phase when it errored out

File: compute_tools/src/compute.rs [hunk 11]
```diff
@@ -816,30 +815,32 @@ impl ComputeNode {
...
-
-                    zenith_admin_connstr
-                        .set_username("zenith_admin")
-                        .map_err(|_| anyhow::anyhow!("invalid connstr"))?;
+                    let mut zenith_admin_conf = postgres::config::Config::from(conf.clone());
+                    zenith_admin_conf.user("zenith_admin");
 
                     let mut client =
-                        Client::connect(zenith_admin_connstr.as_str(), NoTls)
+                        zenith_admin_conf.connect(NoTls)
```
[the remark this answers is not in the document]
Comment (ololobus) [reply compute_tools/src/compute.rs:843]: I've tried doing that, but with tokio-postgres we need to start connection in another thread and here we want to catch gracefully if connection failed. So I failed miserably to refactor it in reasonable time. Will be happy to see a PR that does this properly

File: compute_tools/src/installed_extensions.rs [hunk 12]
```diff
@@ -42,75 +41,51 @@ fn list_dbs(client: &mut Client) -> Result<Vec<String>> {
...
-        .enable_all()
-        .build()
-        .expect("failed to create runtime");
-    let result = rt
-        .block_on(crate::installed_extensions::get_installed_extensions(
-            connstr,
-        ))
-        .expect("failed to get installed extensions");
+    let res = InstalledExtensions {
+        extensions: extensions_map.values().cloned().collect(),
```
[the remark this answers is not in the document]
Comment (ololobus) [reply compute_tools/src/installed_extensions.rs:85]: Didn't want to delay this PR any longer. Also had some follow-ups in mind, so created https://github.com/neondatabase/neon/pull/9934, where applied this change as well

File: compute_tools/src/compute.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ololobus) [reply compute_tools/src/compute.rs:935]: As far as I can see, none of new `apply_*` functions log errors. They are just propagated upper by the stack. So yes, you understand well how instrument works, but it doesn't help here :)

At the end error looks like
```
ERROR could not start the compute node: db error: FATAL: database "what_i_read_today_dev" does not exist: FATAL: database "my_missing_db" does not exist
```
See https://neondb.slack.com/archives/C03H1K0PGKH/p1732732149678949?thread_ts=1732731635.599039&cid=C03H1K0PGKH

We can refactor these new apply functions further and add better errors handling and logging, but it's clearly out of scope for this PR

(let me know if I still miss something)

[the remark this answers is not in the document]
Comment (ololobus) [reply compute_tools/src/compute.rs:935]: Maybe it's possible, btw, just needs more tooling around
